modules/unspents/src/dimensions.ts

```diff
@@ -290,434 +154,392 @@ Object.defineProperty(Dimensions.prototype, 'nInputs', {
- * @return {Dimensions}
- */
-Dimensions.fromTransaction = function ({ ins, outs }, params) {
-  return Dimensions.fromInputs(ins, params).plus(Dimensions.fromOutputs(outs));
-};
+    switch (scriptTypeForChain(chain)) {
+      case 'p2sh':
+        return Dimensions.sum({ nP2shInputs: 1 });
+      case 'p2shP2wsh':
+        return Dimensions.sum({ nP2shP2wshInputs: 1 });
+      case 'p2wsh':
+        return Dimensions.sum({ nP2wshInputs: 1 });
+      case 'p2tr':
+        switch (params.p2trSpendType) {
+          case 'keypath':
+            return Dimensions.sum({ nP2trKeypathInputs: 1 });
+          case 'scriptpath-level1':
+            return Dimensions.sum({ nP2trScriptPathLevel1Inputs: 1 });
+          case 'scriptpath-level2':
+            return Dimensions.sum({ nP2trScriptPathLevel2Inputs: 1 });
+          default:
+            throw new Error(`unsupported p2trSpendType: ${params.p2trSpendType}`);
+        }
+    }
 
-/**
- * @param dimensions (can be partially defined)
- * @return new dimensions with argument added
- */
-Dimensions.prototype.plus = function (dimensions: Partial<IDimensions>) {
-  if (!_.isObject(dimensions)) {
-    throw new TypeError(`expected argument to be object`);
+    throw new Error(`unsupported chain ${chain}`);
   }
 
-  // Catch instances where we try to initialize Dimensions from partial data using deprecated parameters
-  // using only "nOutputs".
-  if ('nOutputs' in dimensions) {
-    if (!('outputs' in dimensions)) {
-      throw new Error('deprecated partial addition: argument has key "nOutputs" but no "outputs"');
+  /**
+   * @param unspents
+   * @return {Dimensions} sum of the dimensions for each unspent (@see Dimensions.fromUnspent())
+   */
+  static fromUnspents(unspents: { chain: ChainCode }[]): Dimensions {
+    if (!Array.isArray(unspents)) {
+      throw new TypeError(`unspents must be array`);
     }
+    // Convert the individual unspents into dimensions and sum them up
+    return Dimensions.sum(...unspents.map((u) => Dimensions.fromUnspent(u)));
+  }
 
-    const { outputs, nOutputs } = dimensions as IDimensions;
+  /**
+   * @param transaction - bitcoin-like transaction
+   * @param [param.assumeUnsigned] - default type for unsigned inputs
+   * @return {Dimensions}
+   */
+  static fromTransaction(
+    {
+      ins,
+      outs,
+    }: {
+      ins: utxolib.TxInput[];
+      outs: utxolib.TxOutput[];
+    },
+    params?: FromInputParams
+  ): Dimensions {
+    return Dimensions.fromInputs(ins, params).plus(Dimensions.fromOutputs(outs));
+  }
 
-    if (outputs.count !== nOutputs) {
-      throw new Error('deprecated partial addition: inconsistent values for "nOutputs" and "outputs.count"');
+  /**
+   * @param dimensions (can be partially defined)
+   * @return new dimensions with argument added
+   */
+  plus(dimensions: Partial<Dimensions>): Dimensions {
+    if (typeof dimensions !== 'object') {
+      throw new TypeError(`expected argument to be object`);
```

Comment:
This also seems like unreachable code assuming the types are respected.

Reply:
It is not reachable from typescript but we still have some JS callsites. I would not add these types of guards in new codes for for now I want to retain them here.